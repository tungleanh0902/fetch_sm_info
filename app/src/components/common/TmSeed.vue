<template>
  <div class="seed-phrase">
    <div class="seed-warning">
      <p>
        This backup code is all that is needed to access your account. Please
        make sure to store it in a safe place.
      </p>
    </div>
    <div
      v-clipboard:copy="value"
      v-clipboard:success="() => onCopy()"
      class="copy-seed"
    >
      Copy to clipboard
      <i
        class="material-icons notranslate copied"
        :class="{ active: copySuccess }"
      >
        check
      </i>
    </div>
    <div>
      <table class="seed-table">
        <tr>
          <td v-for="(word, index) in splitSeed.slice(0, 6)" :key="index">
            <span class="word-number">{{ index + 1 }}</span>
            {{ word }}
          </td>
        </tr>
        <tr>
          <td v-for="(word, index) in splitSeed.slice(6, 12)" :key="index">
            <span class="word-number">{{ index + 7 }}</span>
            {{ word }}
          </td>
        </tr>
        <tr>
          <td v-for="(word, index) in splitSeed.slice(12, 18)" :key="index">
            <span class="word-number">{{ index + 13 }}</span>
            {{ word }}
          </td>
        </tr>
        <tr>
          <td v-for="(word, index) in splitSeed.slice(18, 24)" :key="index">
            <span class="word-number">{{ index + 19 }}</span>
            {{ word }}
          </td>
        </tr>
      </table>
    </div>
  </div>
</template>

<script>
export default {
  name: `TmSeed`,
  props: {
    value: {
      type: String,
      default: ``,
    },
  },
  data: () => ({
    copySuccess: false,
  }),
  computed: {
    splitSeed: function () {
      return this.value.split(` `)
    },
  },
  methods: {
    onCopy() {
      this.copySuccess = true
      setTimeout(() => {
        this.copySuccess = false
      }, 2500)
    },
  },
}
</script>
<style scoped>
.seed-table {
  width: calc(100% + 8px);
  border-spacing: 4px;
  border-collapse: separate;
  margin: 0 -4px;
}

.seed-table td {
  text-align: center;
  width: 16.6666666666667%;
  background-color: var(--app-fg);
  padding: 0.5rem 0.1rem 0.5rem 0.1rem;
  border-radius: 0.2rem;
  font-size: 0.9rem;
  color: var(--bright);
}

.seed-table td .word-number {
  display: block;
  width: 100%;
  opacity: 0.5;

  /* Prevent user to copy word numbers, we only want the words in their correct order */
  -moz-user-select: none;
  -webkit-user-select: none;
  -ms-user-select: none;
  -o-user-select: none;
  user-select: none;
}

.copy-seed {
  display: initial;
  float: right;
  font-size: 0.8rem;
  cursor: pointer;
  margin-bottom: 0.2rem;
  color: var(--link);
}

.copy-seed .material-icons {
  font-size: 12px;
}

.seed-phrase .copied {
  padding-bottom: 2px;
  padding-right: 0;
  transition: opacity 500ms ease;
  color: var(--success);
  opacity: 0;
}

.seed-phrase .copied.active {
  opacity: 1;
}

.seed-warning {
  border: 2px solid var(--warning);
  border-radius: 0.25rem;
  margin-bottom: 1rem;
  padding: 0.5rem;
}

.seed-warning p {
  font-size: var(--sm);
  color: var(--warning);
  margin: 0;
}

@media screen and (max-width: 360px) {
  .seed-table td {
    padding: 0.5rem 0.1rem 0.5rem 0.1rem;
    border-radius: 0.2rem;
    font-size: 0.7rem;
  }
}

@media screen and (min-width: 361px) and (max-width: 667px) {
  .seed-table td {
    font-size: 0.8rem;
  }
}
</style>
