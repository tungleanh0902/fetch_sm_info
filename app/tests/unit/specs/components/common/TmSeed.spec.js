import { shallowMount, createLocalVue } from "@vue/test-utils"
import VueClipboard from "vue-clipboard2"
import TmSeed from "src/components/common/TmSeed"

const localVue = createLocalVue()
localVue.directive(`clipboard`, VueClipboard)

describe(`TmSeed`, () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(TmSeed, {
      localVue,
      propsData: {
        value: `two collect olive inside assault finger relief shallow lottery sugar universe fatigue knock current only absurd famous work path zone distance churn include flower`,
      },
      data: () => ({
        copySuccess: false,
      }),
    })
  })

  it(`should render the seed in a table component`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should split the seed in a 24 words array`, () => {
    const seedWords = [
      `two`,
      `collect`,
      `olive`,
      `inside`,
      `assault`,
      `finger`,
      `relief`,
      `shallow`,
      `lottery`,
      `sugar`,
      `universe`,
      `fatigue`,
      `knock`,
      `current`,
      `only`,
      `absurd`,
      `famous`,
      `work`,
      `path`,
      `zone`,
      `distance`,
      `churn`,
      `include`,
      `flower`,
    ]
    expect(wrapper.vm.splitSeed).toEqual(seedWords)
  })

  describe(`onCopy`, () => {
    it(`should set and reset copySuccess`, () => {
      jest.useFakeTimers()
      wrapper.vm.onCopy() // old test style to make timer work
      expect(wrapper.vm.copySuccess).toBe(true)

      jest.runAllTimers()
      expect(wrapper.vm.copySuccess).toBe(false)
    })
  })

  // TODO: not sure how to test the v-clipboard directive events
  xit(`clicking copy copies the address`, () => {
    jest.useFakeTimers()
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(false)

    wrapper.find(`.address`).trigger(`click`)
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(true)

    jest.runAllTimers()
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(false)
  })
})
