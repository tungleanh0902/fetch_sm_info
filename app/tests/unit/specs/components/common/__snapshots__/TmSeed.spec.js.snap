// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSeed should render the seed in a table component 1`] = `
<div
  class="seed-phrase"
>
  <div
    class="seed-warning"
  >
    <p>
      
      This backup code is all that is needed to access your account. Please
      make sure to store it in a safe place.
    
    </p>
  </div>
   
  <div
    class="copy-seed"
  >
    
    Copy to clipboard
    
    <i
      class="material-icons notranslate copied"
    >
      
      check
    
    </i>
  </div>
   
  <div>
    <table
      class="seed-table"
    >
      <tr>
        <td>
          <span
            class="word-number"
          >
            1
          </span>
          
          two
        
        </td>
        <td>
          <span
            class="word-number"
          >
            2
          </span>
          
          collect
        
        </td>
        <td>
          <span
            class="word-number"
          >
            3
          </span>
          
          olive
        
        </td>
        <td>
          <span
            class="word-number"
          >
            4
          </span>
          
          inside
        
        </td>
        <td>
          <span
            class="word-number"
          >
            5
          </span>
          
          assault
        
        </td>
        <td>
          <span
            class="word-number"
          >
            6
          </span>
          
          finger
        
        </td>
      </tr>
       
      <tr>
        <td>
          <span
            class="word-number"
          >
            7
          </span>
          
          relief
        
        </td>
        <td>
          <span
            class="word-number"
          >
            8
          </span>
          
          shallow
        
        </td>
        <td>
          <span
            class="word-number"
          >
            9
          </span>
          
          lottery
        
        </td>
        <td>
          <span
            class="word-number"
          >
            10
          </span>
          
          sugar
        
        </td>
        <td>
          <span
            class="word-number"
          >
            11
          </span>
          
          universe
        
        </td>
        <td>
          <span
            class="word-number"
          >
            12
          </span>
          
          fatigue
        
        </td>
      </tr>
       
      <tr>
        <td>
          <span
            class="word-number"
          >
            13
          </span>
          
          knock
        
        </td>
        <td>
          <span
            class="word-number"
          >
            14
          </span>
          
          current
        
        </td>
        <td>
          <span
            class="word-number"
          >
            15
          </span>
          
          only
        
        </td>
        <td>
          <span
            class="word-number"
          >
            16
          </span>
          
          absurd
        
        </td>
        <td>
          <span
            class="word-number"
          >
            17
          </span>
          
          famous
        
        </td>
        <td>
          <span
            class="word-number"
          >
            18
          </span>
          
          work
        
        </td>
      </tr>
       
      <tr>
        <td>
          <span
            class="word-number"
          >
            19
          </span>
          
          path
        
        </td>
        <td>
          <span
            class="word-number"
          >
            20
          </span>
          
          zone
        
        </td>
        <td>
          <span
            class="word-number"
          >
            21
          </span>
          
          distance
        
        </td>
        <td>
          <span
            class="word-number"
          >
            22
          </span>
          
          churn
        
        </td>
        <td>
          <span
            class="word-number"
          >
            23
          </span>
          
          include
        
        </td>
        <td>
          <span
            class="word-number"
          >
            24
          </span>
          
          flower
        
        </td>
      </tr>
    </table>
  </div>
</div>
`;
